### PR TITLE
[windows] support -addon in build script

### LINKED ISSUE #5987: Windows installer should not embed mingw make
When building math-comp on windows I made the following steps:
```bash
# export PATH=/cygdrive/c/Coq64/bin
# make
```
This failed in ways that are hard to imagine (i.e. even the way the win shell chops the command line into arguments is different..).

The good news is that once I run `rm /cygdrive/c/Coq64/bin/make.exe` then `make` did just work.

Given that such `make.exe` binary does not run the makefiles generated by `coq_makefile`, well... I guess we could just not ship it.

### PR DESCRIPTION
**Pull request by gares**
This PR adds a very basic support for addons to be built after Coq and installed inside Coq (eg in user-contribs).

The only addon used here is bignums, to fix the regression introduced in 8.7 (harder to build pure .v libraries that use bignum on windows).

@maximedenes when this is merged a blocker issue has to be opened: we download and install bignums *master*, since there is no tag compiling on 8.8~alpha. I guess one has to release bignums (tag) and update the build script with the desired version.

The best reviewer is @MSoegtropIMC 

Fixes #5987

### CODE REVIEW

File: dev/build/windows/makecoq_mingw.sh [hunk 1]
```diff
@@ -1320,6 +1326,22 @@ function make_coq_installer {
...
+  if build_prep https://github.com/coq/bignums/archive/ master zip 1 bignums-8.8.0; then
+    logn make make all
+    logn make-install make install
+    build_post
+  fi
+}
+
+function make_addons {
+  if [ ! -z "$COQ_ADDONS" ]; then
+    for addon in $COQ_ADDONS; do make_addon_$addon; done
```
Comment (SkySkimmer) [inline dev/build/windows/makecoq_mingw.sh]: Why test for nonzeroness? A loop over nothing does nothing.

Comment (gares) [reply dev/build/windows/makecoq_mingw.sh]: oops, thanks

**Comment by MSoegtropIMC** [comment]
Dear Enrico,

It wouldn't be difficult to support a separate installer group for this. The build scripts have support for this. At the end there are two lines:

list_files ocaml
list_files ocaml_coq

These do a directory tree of the install target file system and the files to be installed by the installer in different groups are taken from diffs of these file lists. The logic is in make_coq_installer and there are convenient shell functions available for the Boolean ops needed.

To add a section "plugins" you would do another list_files, add the corresponding logic to make_coq_installer and add a section to the NSIS file.

Otherwise the changes are fine.

Btw.: what I intended to do is a bit different: a script to setup an environment which allows to compile and install plugins on the command line, rather than to compile plugins with the installer build scripts.

Best regards,

Michael

**Review by MSoegtropIMC — APPROVED** [review]
Two comments:

1.) I would prefer a separate installer section as outlined in my previous comment, but don't see this as mandatory, more nice to have,

2.) Editorial: the spacing is not aligned in ECHO -addon   = %COQ_ADDONS% in the .bat file.

**Comment by gares** [comment]
I tried to address all issues, but I could not test thing properly in my local setup, so let's see what appveyor says

**Comment by gares** [comment]
On a side note, I think the "files for plugin developers" business is not 100% correct.
For example `.cma` files (in `plugins/`) are used (in place of `cmxs`) if Coq *byte* is executed, but they seem to be in the pluginsdev package.

**Comment by gares** [comment]
For @maximedenes : the last commit changes the installer. If it does not work one can merge the other 3 commits and still fix the regression (in case we are in a hurry for the beta release).

**Comment by Zimmi48** [comment]
@gares If the last commit generates a separate installer you need to update `appveyor.yml` to inform it about this new artifact.

**Comment by gares** [comment]
It's a separate "packge" for the same installer (1 file, but lets you select among 3 packages, rather than 2)

**Comment by gares** [comment]
Since yesterday I had this error. Here I managed to build the installer by using a local copy of NSIS.
```
[00:37:12] Download failed: http://downloads.sourceforge.net/project/nsis/NSIS%202/2.51/nsis-2.51.zip
[00:37:12] + echo The file wget downloaded is an html file:
[00:37:12] The file wget downloaded is an html file:
[00:37:12] + cat nsis-2.51.zip
[00:37:12] <html><head>
[00:37:12] <title>SourceForge</title>
[00:37:12] <!-- <script src="/js/jquery.com/jquery-1.11.0.min.js"></script> -->
[00:37:12] <script src="https://code.jquery.com/jquery-1.11.0.min.js"></script>
[00:37:12] <script src="https://sourceforge.net/js/mirrors.js"></script>
[00:37:12] <script src="/js/sf.js"></script>
[00:37:12] <script>
[00:37:12] var DR_loc = DR_parse_hash_url();
[00:37:12] if (DR_loc) {
[00:37:12]     DR_sf_main(DR_loc);
[00:37:12] } else {
[00:37:12]     window.location.href = 'https://sourceforge.net/home.html';
[00:37:12] }
[00:37:12] </script>
[00:37:12] </head><body>
[00:37:12] <noscript>
[00:37:12] We're sorry -- the Sourceforge site is currently in Disaster Recovery mode, and currently requires
[00:37:12] the use of javascript to function.  Please check back later.
[00:37:12] </noscript>
[00:37:12] </body></html>
[00:37:12] + exit 1
```
A possible workaround is to hardcode a mirror name in the download URL of NSIS (or any other thing hosted by SF).

**Comment by Zimmi48** [comment]
FTR I can also see this message when going to sourceforge.net because of my NoScript extension.

Anyways, there's a high rate of failures with these packaging scripts due to the use of unreliable download locations. I think it would be good to have a generic solution for this. Mirroring on our own servers or something along these lines. See also #6758 and #6520.

**Comment by MSoegtropIMC** [comment]
I prefer to download from "official sites" (to document where things come from originally). Of cause it would make sense to add support for a fallback mirror location. But I would report download failures from the primary location as warning.
One of the reasons I did these scripts was that the original Windows build did quite a few downloads from rather obscure sources, and I want to avoid falling back to this long term.

**Comment by MSoegtropIMC** [comment]
Regarding make: I agree that it should be removed until I fixed make and OCaml, but wouldn't it make more sense to do this by setting the corresponding build flag to N?

Best regards,
Michael

**Comment by MSoegtropIMC** [comment]
Regarding a fallback download location: do you have something suitable at INRIA? Otherwise I could upload all required files on one of my servers and implement the fallback.

Total size for everything except Cygwin binary packages (but including Cygwin sources) is about 1GB (including a few old versions). Cygwin binary packages are another 1.3 GB, so total is < 2.5 GB.

If it is not used as primary location, the traffic should also be negligible.

**Comment by Zimmi48** [comment]
There is another problem some of the current download sources: some are HTTP and not HTTPS. This means that our package build scripts are potentially vulnerable to some attacks.

> Regarding a fallback download location: do you have something suitable at INRIA? Otherwise I could upload all required files on one of my servers and implement the fallback.

@letouzey or @maximedenes might want to answer that.

**Comment by maximedenes** [comment]
I can put it online in a subfolder of the coq website.

**Comment by MSoegtropIMC** [comment]
@Théo: Indeed, the http issue should be fixed or a hash check should be done (the hashes can be embedded into the build script if they are not otherwise available). Please report a bug and assign it to me.

@Maxime: the build scripts maintain a cache folder for Cygwin and for other downloads (called source cache, but it also contains a few binary downloads). These two folders contain all required files. Do you want to cache Cygwin as well? If so, I will find out what folder structure the cygwin setup needs.

**Comment by MSoegtropIMC** [comment]
P.S.: sorry for referencing unrelated people (maxime and th) - in case you get mails about this please ignore and take my apologizes.

Best regards,
Michael

**Comment by gares** [comment]
I don't know appveyor well enough, but if it has a cache (as ci/travis) we may as well use it to store the `source_cache/` directory (maybe we already do, I really don't know)

**Comment by gares** [comment]
I've fixed the make thing as you say.

FTR: -make is incompatible with -addon becase:
- `coq_inst/bin` is *prefixed* to `PATH`, hence takes precedence
- the makefiles of coq packages (bignums for example) don't seem to work with that flavor of make

**Comment by gares** [comment]
This PR superseeds #5991, and I'm now closing it.

**Comment by Zimmi48** [comment]
Please keep the reference to #5987 in the commit message or the PR description so that the issue is automatically closed on merging.

**Comment by maximedenes** [comment]
We need to see if the appveyor job is fixed to test this PR.

**Comment by gares** [comment]
This time it should build the 32 bit installer too.
The issue is always the same, command lines too long. I hack around it.

**Comment by gares** [comment]
Yes: https://ci.appveyor.com/project/gares/coq it works on 32 bit windows.

**Comment by gares** [comment]
@maximedenes this is not ready for merge.
I've opened #6929 to track the fact that we are shipping bignums master rather than a tag.

**Comment by gares** [comment]
Ops, i meant NOW ready for merge. I believe it is more important to have the script in place than having the tag.

 Also the tag has to be made after the freeze, when no API change take place, so it is a chicken egg problem I solved by using a sloppy URL and request a one line change. Sure you can solve it by giving this pr an exception too, but I prefer the converse

**Comment by MSoegtropIMC** [comment]
Hi Enrico,

can you tell a bit more about the "command line too long" issue? At which part of the build does it happen?

Best regards,

Michael

**Comment by gares** [comment]
Here it is:
1. `coq_makefile` puts in `Makefile.coq.conf` the list of direcotries in which Coq's .cmi files are
1. `Makefile.coq` prefixes these directories with `-I ` and calls `ocamlfind opt ... -I .... `
1. passing -verbose to `ocamlfind` I discovered it calls `cygpath -m <stuff>`  where `stuff` contains the list of `-I` plus a bunch of other things I don't really understand
1. the error is raised (on win32 only, win64 is fine)

We had similar problems before.
The (most likely) general solution is to switch to ocaml 4.05 and use `-arg FileWithAllTheArguments`.
In this specific case it is `ocamlfind` that calls `cygpath` with such a long list of argument, so maybe `-arg` is not going to solve this issue.

The hack simply prunes some `-I..` that are not needed by `bignums`

**Comment by MSoegtropIMC** [comment]
Dear Enrico,

I see. This should also be fixed by the replacement for OCaml system command I am working on (which is in many other ways more compatible to Unix).

Best regards,

Michael

**Pull request merged**